src/LanguageServer/Impl/Implementation/WorkspaceManager.cs

```diff
@@ -0,0 +1,74 @@
+﻿using Microsoft.Extensions.FileSystemGlobbing;
+using Microsoft.Extensions.FileSystemGlobbing.Abstractions;
+using Microsoft.PythonTools.Analysis;
+using Microsoft.PythonTools.Analysis.Infrastructure;
+using Microsoft.PythonTools.Parsing;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Python.LanguageServer.Implementation {
+    class WorkspaceManager {
+        private string _rootDir;
+        private string[] _incluideFiles;
+        private string[] _excludeFiles;
+        private IndexingFileParser _indexingFileParser;
+        private Matcher _matcher;
+
+        public WorkspaceManager(string rootDir, string[] includeFiles, string[] excludeFiles, IndexingFileParser indexingFileParser) {
+            _rootDir = rootDir;
+            _incluideFiles = includeFiles;
+            _excludeFiles = excludeFiles;
+            _indexingFileParser = indexingFileParser;
+            if (string.IsNullOrEmpty(_rootDir)) {
+                throw new ArgumentException($"{nameof(rootDir)} null or empty", nameof(rootDir));
+            }
+            InitializeMatcher();
+        }
+
+        private Task<IEnumerable<string>> AsyncDirectoryFilePaths() {
+            return Task.Run(() => {
+                return DirectoryFilePaths();
+            });
+        }
+
+        private void InitializeMatcher() {
+            _matcher = new Matcher();
+            _matcher.AddIncludePatterns(_incluideFiles.IsNullOrEmpty() ? new[] { "**/*" } : _incluideFiles);
+            _matcher.AddExcludePatterns(_excludeFiles ?? Enumerable.Empty<string>());
+        }
+
+        private IEnumerable<string> DirectoryFilePaths() {
+
+            var dib = new DirectoryInfoWrapper(new DirectoryInfo(_rootDir));
+            var matchResult = _matcher.Execute(dib);
+            foreach (var file in matchResult.Files) {
+                var path = Path.Combine(_rootDir, PathUtils.NormalizePath(file.Path));
+                if (ModulePath.IsPythonSourceFile(path)) {
+                    yield return path;
+                }
+            }
+        }
+
+        public async void ParseRootDirAsync() {
+            foreach (var path in await AsyncDirectoryFilePaths()) {
+                _indexingFileParser.ParseForIndex(path);
+            }
+        }
+
+        public void StartParseRootDir() {
+            Task.Run(() => {
+                ParseRootDirAsync();
+            }).Start();
```

Comment:
`Run` starts the task. Oh, and void starting task is not a good idea. What happens if we want to cancel it.